src/hardware/mixer.cpp

```diff
@@ -368,6 +375,45 @@ void MixerChannel::AddSilence()
 	MIXER_UnlockAudioDevice();
 }
 
+void MixerChannel::EnableLowPassFilter(const bool enabled)
+{
+	filter.enabled = enabled;
+}
+
+void MixerChannel::ForceLowPassFilter(const bool force)
+{
+	filter.force = force;
+}
+
+void MixerChannel::ConfigureLowPassFilter(const uint8_t order,
+                                          const uint16_t cutoff_freq)
+{
+	assert(order > 0 && order <= max_filter_order);
+	const auto sample_rate = mixer.freq;
+	for (auto i = 0; i < 2; ++i)
+		filter.lpf[i].setup(order, sample_rate, cutoff_freq);
+}
+
+void MixerChannel::EnableZeroOrderHoldUpsampler(const bool enabled)
+{
+	zoh_upsampler.enabled = enabled;
+}
+
+void MixerChannel::UpdateZOHUpsamplerState()
+{
+	// we only allow upsampling
+	zoh_upsampler.step = std::min(static_cast<float>(sample_rate) /
+	                                      zoh_upsampler.target_freq,
+	                              1.0f);
+}
```

Comment:
If a step of `1.0f` means the sampler is grinding through the math but effectively not changing the rate, I wonder if this function could behave similar to the auto-sleep logic in the channels?

``` c++
	const auto source_freq = static_cast<float>(sample_rate);
	const auto & target_freq = zoh_upsampler.target_freq;

	if (target_freq <= source_freq + FLT_EPSILON) {
		ToggleZeroOrderHoldUpsampler(FilterState::Off);
		LOG_INFO("FILTER: ... message about rates and putting the filter to sleep");
	}
	assert(target_freq > source_freq);
	zoh_upsampler.step = source_freq / target_freq;
}
```

Reply:
Ah yeah, I just slapped the `min` stuff on top of everything, thinking that would do the job -- which it does, but it's better to turn off the upsampler in this case, given we have the ability to do so.

Sleep wouldn't make much sense in this case because it's either on or off; it cannot be re-enabled after a while based on some condition (the DAC rate is constant, the host rate is also constant in the sense that it doesn't change unexpectedly at runtime).

Reply:
Thanks  - all makes sense.

Reply:
You know what, after thinking a bit more about it, I'll just leave it as it is. The `min` would only clamp `step` if the target rate is above 45454 -- it should never happen as that's the max DAC rate of the SB. So it's just a safety measure, we might as well just leave it off, but it's better to clamp, "just in case".

I don't think any games use that rate anyway. And if they do, well, the upsampler will in effect do nothing, and it's so efficient it doesn't matter. It's just it would complicate the code more to disable the resampler when setting the frequency, then `EnableZeroOrderHoldUpsampler` would need to take that into account -- I see it as a complication with no practical benefits.

Reply:
Thanks for sharing your thinking, and I fully agree.